packages/test/test-end-to-end-tests/src/test/benchmark/opCriticalPath.time.spec.ts

```diff
@@ -0,0 +1,125 @@
+			},
+		},
+	},
+};
+
+type Patch<T, U> = Omit<T, keyof U> & U;
+
+type ContainerRuntime_WithPrivates = Patch<ContainerRuntime, { flush: () => void }>;
+
+describeCompat(
+	"Op Critical Paths - runtime benchmarks",
+	"NoCompat",
+	(getTestObjectProvider) => {
+		let provider: ITestObjectProvider;
+		let mainContainer: IContainer;
+		let defaultDataStore: ITestDataObject;
+		let containerRuntime: ContainerRuntime_WithPrivates;
+
+		const before = async () => {
+			provider = getTestObjectProvider();
+			const loader = provider.makeTestLoader(testContainerConfig);
+			mainContainer = await loader.createDetachedContainer(provider.defaultCodeDetails);
+
+			await mainContainer.attach(provider.driver.createCreateNewRequest());
+			defaultDataStore = (await mainContainer.getEntryPoint()) as ITestDataObject;
+			containerRuntime = defaultDataStore._context
+				.containerRuntime as ContainerRuntime_WithPrivates;
+
+			defaultDataStore._root.set("force", "write connection");
+			await provider.ensureSynchronized();
+		};
+		function sendOps(label: string) {
+			Array.from({ length: 100 }).forEach((_, i) => {
+				defaultDataStore._root.set(`key-${i}`, `value-${label}`);
+			});
+
+			containerRuntime.flush();
+		}
+
+		benchmark({
+			title: "Submit+Flush",
+			before, // Set up container for a write connection
+			benchmarkFnAsync: async () => {
+				sendOps("A");
+				const opsSent = await timeoutPromise<number>(
+					(resolve) => {
+						containerRuntime.deltaManager.outbound.once("idle", resolve);
+					},
+					{ errorMsg: "container2 outbound queue never reached idle state" },
+				);
+				assert(opsSent > 0, "Expecting op(s) to be sent.");
+			},
+		});
+	},
+);
+
+describeCompat(
+	"Op Critical Paths - runtime benchmarks",
+	"NoCompat",
+	(getTestObjectProvider) => {
+		let provider: ITestObjectProvider;
+		let mainContainer: IContainer;
+		let defaultDataStore: ITestDataObject;
+		let containerRuntime: ContainerRuntime_WithPrivates;
+
+		const before = async () => {
+			provider = getTestObjectProvider();
+			const loader = provider.makeTestLoader(testContainerConfig);
+			mainContainer = await loader.createDetachedContainer(provider.defaultCodeDetails);
+
+			await mainContainer.attach(provider.driver.createCreateNewRequest());
+			defaultDataStore = (await mainContainer.getEntryPoint()) as ITestDataObject;
+			containerRuntime = defaultDataStore._context
+				.containerRuntime as ContainerRuntime_WithPrivates;
+
+			defaultDataStore._root.set("force", "write connection");
+			await provider.ensureSynchronized();
+		};
+
+		function sendOps(label: string) {
+			Array.from({ length: 100 }).forEach((_, i) => {
+				defaultDataStore._root.set(`key-${i}`, `value-${label}`);
+			});
+
+			containerRuntime.flush();
+		}
+
+		benchmark({
+			title: "Roundtrip",
+			before, // Set up container for a write connection
+			benchmarkFnAsync: async () => {
+				sendOps("B");
```

Comment:
What I wish I could do - but couldn't figure out how - is to have this `sendOps` line happen outside the measurement _on each iteration_.  Is it possible?  As-is I will just subtract the time from the other benchmark if I want to think about process time only.  But it's not quite right because it includes local server sequencing time too (I wanted to just resume the inbound queue - all the ops would be sitting there)

Reply:
I don't think it's possible today. There's `beforeEachBatch` in the property bag we pass to `benchmark()`, but that still doesn't do what you want here. Quick context (only applies to runtime benchmarks, not memory or custom ones): since some things we want to measure take so little time (pure functions might take nanoseconds) and system clock resolution might not be good enough for accurate measurements at that scale, the way benchmarking works is that it first runs the function to benchmark many times to determine how many iterations in a row are enough to get a good measurement given the system clock resolution, and then runs batches with that same number of iterations each time, and returns the average runtime for each batch as one data point. Many batches are run, and the final result is the average of those. `beforeEachBatch` lets you do something before each batch runs, but not before each call to `benchmarkFnAsync` inside the batch.

All that said... given what we're testing here, maybe we can make it work, or it might even work without further tweaks. Thing start [by assuming that we should run 1 iteration per batch](https://github.com/microsoft/FluidFramework/blob/37c0f2d1c903f12ba6a83e26de2a664f025794aa/tools/benchmark/src/runBenchmark.ts#L120). Then [we increment that number during the "`AdjustIterationPerBatch` phase" if running a batch took less time than a configured minimum](https://github.com/microsoft/FluidFramework/blob/37c0f2d1c903f12ba6a83e26de2a664f025794aa/tools/benchmark/src/runBenchmark.ts#L148-L152), which if I'm reading the code correctly is 50ms or could be higher depending on the resolution of the clock in the machine where the benchmarks run. So if the async function under benchmark already takes longer than that, we won't increment the iterations per batch from the default 1, and you can use `beforeEachBatch` for what you want. But we can play it even safer, passing `startPhase: Phase.CollectData` so we completely skip the `AdjustIterationsPerBatch` phase and are guaranteed that there's only 1 iteration per batch. The main caveat I see is that if the async function ends up being too quick after that, the measurements might not be accurate.

cc @CraigMacomber in case he has thoughts, he's more familiar than I am with this code and how it works.

Reply:
Would you not be able to do it in the `before` function? As in, pause the inbound queue and send the messages (most likely need another mechanism to make sure ops are received back as I think `ensureSynchronized` waits until after processed)

Reply:
As discussed offline with Alex - `beforeEachBatch` doesn't support `async` setup code, so it doesn't help separate the submit/sequence time for me.  Opening a follow-up task to use it if/when that is changed to be `async`.

Reply:
AB#19813 and AB#19814